### ISSUE TITLE
dcr: BotStateManager should be split into separate ConversationState and UserState middleware

### ISSUE THREAD

**Issue by Stevenic** [Microsoft/botbuilder-dotnet#156]
Currently we have a single `BotStateManager` middleware that's responsible for reading & writing both conversation state and user state. I'd like to propose that we replace this with two separate pieces of middleware that have logic for managing conversation state and user state independently. The reasoning for that is:

- You may want to use one but not the other and while we have config switches to turn off either piece of state it would be clearer if they were separate middleware.
- You may want to use a `RedisStorage` provider for storing conversation state and a separate `AzureTableStorage` provider for user state. Again, we can handle this through configuration but it's cleaner if they're separate middleware.
- Tracking conversation state needs slightly different semantics then tracking user state. This is the bigger reason for separating the two. With conversation state if you see an "endOfConversation" activity sent or received you should delete the current conversation state. This is needed to prevent Conrtana skills from getting into a stuck state but is generally just a good way to help your bot avoid stuck states.
- Key value storage is fine for tracking conversation state but user state is probably better tracked as a graph so in production systems it seems likely that you'd rarely want our key-value based user state manager.

The con to splitting these up is that should you want to run both you now have to read them in series. I have another proposal to create a `ParallelMiddlewareSet` which would address that as there are plenty of other things you'd want to do in parallel in a production bot.

**Comment by cleemullins** [comment]
Code is in and awaiting review. I'll merge it in once another set of eyes looks at it.

**Comment by Stevenic** [comment]
@cleemullins is rocking :) I have some catching up to do... lol

**Comment by JonathanFingold** [comment]
Sorry to be late to the conversation. I advocate for not making the conversation-user distinction at all, and leaving it up to the developer to manage the scope of what they put into state/storage. I'm guessing that many devs will end up with three or four types of scope for information they are tracking.

**Comment by Stevenic** [comment]
@JonathanFingold I guess that's generally what I'm recommending as well although there's is very specific additional logic you should run for your conversation state. In particular you really need to look for "endOfConversation" activities going out of your bot and you should delete your conversation state anytime you see one.  That's pretty much a requirement for Cortana otherwise your skill simply won't work right and it's generally a good practice for all bots.

Giving developers a single piece of ConversationState middleware that manages loading/saving state and adds the additional check above seems like goodness to me.  You could split it into two separate pieces of middleware but then that just means you always need to remember to add both to every bot.

**Comment by JonathanFingold** [comment]
@Stevenic That makes sense. I should have looked at Chris' PR first.

**Comment by cleemullins** [comment]
Closing, since PR #159 is merged.
